independent-projects/tools/codestarts/codestarts.adoc

```diff
@@ -0,0 +1,220 @@
+Quarkus extensions are able to provide one or more well defined codestarts which will contain the necessary resources and code examples required to get started using it.
+
+== How it works
+
+There are two kinds of codestarts contributing to the generation of a project, the kind where we want to have only one for a project (called Base) and the rest (called Extra).
+
+Base:
+
+* project: The project skeleton (e.g. a Quarkus project)
+* buildtool: The build tool (e.g. Maven, Gradle, Gradle with Kotlin DSL)
+* language: The coding language (e.g. Java, Kotlin, Scala)
+* config: The config type (e.g. yaml, properties)
+
+Extra:
+* tooling: Anything that can be added to improve the project (e.g. dockerfiles)
+* example: Any Quarkus extension can provide example code. The user can decide to activate it or not.
+
+Each codestart consists of:
+
+. A specific directory
+. A `codestart.yml` file
+. Optionally some templates which follow a structure and naming conventions
+
+*NOTE* The `codestart.yml` file and the directory structure follow the same principle, it can optionally contain a base and/or some language overrides.
+
+=== Project generation
+
+When generating a Quarkus project:
+
+. Codestarts to use are resolved depending on the input
+. The Codestarts' shared data is processed to make it available for all codestarts
+. The relevant files are processed:
+** Only the files that are related to the selected language are processed
+** The files are processed differently based on a naming convention.
+** The data is used to render Qute templates
+** We always process in this order: language, project, buildtool, config, custom, other.
+
+The data used to generate a specific codestart is a merge of:
+
+. The data of the codestart to generate
+. All codestarts "shared" data
+. The user input
+. Some specific post processing (e.g. adding dependencies)
+
+*NOTE* The data (shared or not) can also be specific to a language.
+
+=== Directory Structure
+
+*NOTE* `codestart.yml` is the only required file.
+
+* `codestart.yml` must be at the root of the codestart
+* `./base` contains all the files that will be processed
+* `./[languageName]` contains all the files that will be processed if the specified language has been selected (overriding base)
+
+=== codestart.yml
+
+codestart.yml:
+[source,yaml]
+----
+name: the unique name - REQUIRED
+ref: the reference name (not unique) to use for extension matching (else the name is used)
+type: the type [example (default), project, buildtool, language, config] or any custom type
+fallback: flag to indicate that it should be selected as fallback when no codestart has been selected with type - Only for base types
+preselected: flag to indicate that it should be pre-selected - Only for extra types.
+spec:
+  [base or language name]: the specification (base, java, ...)
+    data: a map of data to use only for this codestart
+    shared-data: a map of data to use accross all codestarts
```

Comment:
we should somewhere list the default name/data codestart properties/data will be fed in.

Reply:
I am preparing a command to display merged `shared-data` for a given `CodestartInput`.

Reply:
aren't there some that are always the same ?

Reply:
yes you have: https://github.com/ia3andy/quarkus/blob/codestarts-v0/independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/CodestartData.java#L19

and also some auto-generated ones:
https://github.com/ia3andy/quarkus/blob/codestarts-v0/independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/CodestartData.java#L90

`codestart-project.<basetype>.name` (ex: `codestart-project.buildtool.name`)

Reply:
https://github.com/quarkusio/quarkus/issues/11085